### PR TITLE
fix: refactor how reachability timeouts work

### PR DESCRIPTION
**Pull request by marcin-bazyl**
# COMPLETES #[SPARK-540198](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-540198)

## This pull request addresses
Currently amplitude metrics show that there are about 3k successful meeting joins from users that failed all reachability checks. Web app wants to show a UI warning to the user when reachability fails, so we need to make sure that reachability results are reliable and when reachability says that all checks failed, it is indeed the case that all clusters are not reachable.

The suspicion is that false negative results are caused by the very short (3s) timeout used for reachability checks.

## by making the following changes

Reorganizing how reachability times out - instead of each ClusterReachability instance having its own timeout, the timeout is managed by the Reachability class. We keep existing behavior of gatherReachability() resolving at most after 3s, but we will now keep collecting the results in the background after that and eventually send metrics with the stats - this will allow us to make a decision on how long the timeouts need to be.

more details here:
https://confluence-eng-gpk2.cisco.com/conf/display/WTWC/Change+of+reachability+timeouts

### Change Type

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested

unit tests

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request

- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [x] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/@webex/plugin-meetings/src/reachability/clusterReachability.ts [hunk 1]
```diff
@@ -211,22 +263,43 @@ export class ClusterReachability {
...
       result.latencyInMilliseconds = latency;
       result.result = 'reachable';
+      if (publicIp) {
+        result.clientMediaIPs = [publicIp];
+      }
+
+      this.emit(
+        {
+          file: 'clusterReachability',
+          function: 'registerIceCandidateListener',
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/reachability/clusterReachability.ts:293]: i think this should be `saveResult` no?

File: packages/@webex/plugin-meetings/src/reachability/index.ts [hunk 2]
```diff
@@ -395,16 +409,196 @@ export default class Reachability {
...
+    });
+  }
+
+  /**
+   * Sends a metric with all the statistics about how long reachability took
+   *
+   * @returns {void}
+   */
+  private async sendMetric() {
+    const getStatistics = (results, protocol, isVideoMesh) => {
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/reachability/index.ts:465]: nitpick: might be better to extract as separate func, that way testing is easier

File: packages/@webex/plugin-meetings/src/reachability/index.ts [hunk 3]
```diff
@@ -395,16 +409,196 @@ export default class Reachability {
...
+
+        this.resolveReachabilityPromise();
+      }
+    }, VIDEO_MESH_TIMEOUT * 1000);
+
+    this.publicCloudTimer = setTimeout(() => {
+      this.publicCloudTimer = undefined;
+
+      LoggerProxy.logger.log(
+        `Reachability:index#gatherReachability --> Reachability checks timed out (${DEFAULT_TIMEOUT}s)`
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/reachability/index.ts:532]: should be `index#startTimers`

File: packages/@webex/plugin-meetings/src/reachability/index.ts [hunk 4]
```diff
@@ -395,16 +409,196 @@ export default class Reachability {
...
+   * @returns {void}
+   */
+  private startTimers() {
+    this.vmnTimer = setTimeout(() => {
+      this.vmnTimer = undefined;
+      // if we are only missing VMN results, then we don't want to wait for them any longer
+      // as they are likely to fail if users are not on corporate network
+      if (this.areAllPublicClusterResultsReady()) {
+        LoggerProxy.logger.log(
+          'Reachability:index#gatherReachability --> Reachability checks timed out (VMN timeout)'
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/reachability/index.ts:521]: should be `index#startTimers`

File: packages/@webex/plugin-meetings/src/reachability/index.ts [hunk 5]
```diff
@@ -395,16 +409,196 @@ export default class Reachability {
...
+          file: 'reachability',
+          function: 'overallTimer timeout',
+        },
+        'reachability:done',
+        {}
+      );
+      this.sendMetric();
+
+      LoggerProxy.logger.log(
+        `Reachability:index#gatherReachability --> Reachability checks fully timed out (${OVERALL_TIMEOUT}s)`
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/reachability/index.ts:553]: should be `index#startTimers`

**Review by torgeadelin — COMMENTED** [review]
some small comments, but haven't had a chance to test, will continue tomorrow

File: packages/@webex/plugin-meetings/src/reachability/index.ts (hunk 5 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/reachability/index.ts]: fixed

File: packages/@webex/plugin-meetings/src/reachability/index.ts (hunk 4 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/reachability/index.ts]: done

File: packages/@webex/plugin-meetings/src/reachability/index.ts (hunk 3 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/reachability/index.ts]: done

File: packages/@webex/plugin-meetings/src/reachability/clusterReachability.ts (hunk 1 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/reachability/clusterReachability.ts]: done

File: packages/@webex/plugin-meetings/src/reachability/index.ts (hunk 2 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/reachability/index.ts]: Ok, I've done that, but both `getStatistics()` and `sendMetric()` are private so they shouldn't be called from the tests anyway. I've worked around this by declaring them as `protected` so in the test I created a class that extends `Reachability` class and added public methods that call `getStatistics()` and `sendMetric()` to allow us to test them. I've added more unit tests for these methods now, I agree it makes it easier to test all the edge cases, but I feel that it makes the unit tests too closely coupled to the implementation, which then makes the tests not very useful if you try to refactor something inside `Reachability` class, so it's a trade off....

Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/reachability/index.ts]: I missed that it was a private function. 
Yeah, I see your point and I agree, but the effort for still doing it is very appreciated!

**Review by torgeadelin — APPROVED** [review]
looking good, tested common join flows and all working good

File: packages/@webex/plugin-meetings/src/reachability/clusterReachability.ts [hunk 6]
```diff
@@ -23,10 +21,32 @@ export type ClusterReachabilityResult = {
   xtls: TransportResult;
 };
 
+// data for the Events.resultReady event
+export type ResultEventData = {
+  protocol: 'udp' | 'tcp' | 'xtls';
+  result: 'reachable' | 'unreachable' | 'untested';
```
Comment (rarajes2) [inline packages/@webex/plugin-meetings/src/reachability/clusterReachability.ts:27]: Q. From where are we triggering the partial reachability event?

File: packages/@webex/plugin-meetings/src/reachability/clusterReachability.ts [hunk 7]
```diff
@@ -211,22 +263,43 @@ export class ClusterReachability {
...
     if (result.latencyInMilliseconds === undefined) {
       LoggerProxy.logger.log(
         // @ts-ignore
-        `Reachability:index#storeLatencyResult --> Successfully reached ${this.name} over ${protocol}: ${latency}ms`
+        `Reachability:index#saveResult --> Successfully reached ${this.name} over ${protocol}: ${latency}ms`
       );
       result.latencyInMilliseconds = latency;
       result.result = 'reachable';
+      if (publicIp) {
+        result.clientMediaIPs = [publicIp];
```
Comment (rarajes2) [inline packages/@webex/plugin-meetings/src/reachability/clusterReachability.ts:287]: Do we need to append the `publicIp` to `clientMediaIPs` here or is this intentional?

File: packages/@webex/plugin-meetings/src/reachability/index.ts [hunk 8]
```diff
@@ -3,19 +3,26 @@
  */
 
 /* eslint-disable class-methods-use-this */
-import {mapValues} from 'lodash';
+import {isEqual, mapValues, mean} from 'lodash';
```
Comment (rarajes2) [inline packages/@webex/plugin-meetings/src/reachability/index.ts:6]: Should we import these methods individually? Not sure if tree shaking works this way

e.g.
```
import isEqual from 'lodash/isEqual';
```

File: packages/@webex/plugin-meetings/src/reachability/index.ts [hunk 9]
```diff
@@ -395,16 +409,207 @@ export default class Reachability {
...
+    this.resultsCount.public.xtls = 0;
+  }
+
   /**
    * Performs reachability checks for all clusters
    * @param {ClusterList} clusterList
-   * @returns {Promise<ReachabilityResults>} reachability check results
+   * @returns {Promise<void>} promise that's resolved as soon as the checks are started
    */
-  private async performReachabilityChecks(clusterList: ClusterList): Promise<ReachabilityResults> {
```
Comment (rarajes2) [inline packages/@webex/plugin-meetings/src/reachability/index.ts:403]: Why did we remove the return type?

File: packages/@webex/plugin-meetings/src/reachability/clusterReachability.ts (hunk 6 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/reachability/clusterReachability.ts:27]: I'm not sure what are you asking about, the `ResultEventData ` is used only with "resultReady" event and that event will actually always have result set to "reachable", because it's only emitted when we get an ICE candidate. I could change the type here to just say  `result: 'reachable'.
Or if you're asking about the partial reachability event that was mentioned on the confluence page, then that its emitted from [here](https://github.com/webex/webex-js-sdk/blob/7f573abc7ef45707c945ea9d3c4c9b59fc7b7f5f/packages/%40webex/plugin-meetings/src/reachability/index.ts#L685) but I've renamed the event and called it 'reachability:firstResultAvailable' instead of "reachability:partiallyDone". I feel it's a better name, once the PR is approved I'll update the confluence.

File: packages/@webex/plugin-meetings/src/reachability/clusterReachability.ts (hunk 7 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/reachability/clusterReachability.ts:287]: it's intentional, because this is the first time we're saving the result (see the check in line 279) so there is no array to append to

File: packages/@webex/plugin-meetings/src/reachability/index.ts (hunk 8 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/reachability/index.ts:6]: I haven't checked if it works correctly with tree shaking but there are 108 other existing occurrences of SDK importing methods from lodash this way, so I just kept the same convention and if we think that this is not the right way, we should probably raise a jira to investigate and fix it across the whole SDK separately to this PR.

File: packages/@webex/plugin-meetings/src/reachability/index.ts (hunk 9 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/reachability/index.ts:403]: Previously gatherReachability() was calling performReachabilityChecks() and that was returning all the results that were then stored in localStorage by gatherReachability() in one go. Now we update results in local storage every time a new result comes in through the resultReady event, so gatherReachability() doesn't need to get the results from performReachabilityChecks() anymore and performReachabilityChecks() doesn't wait for all results to be ready before it resolves and only gatherReachability() waits for that. Also performReachabilityChecks() is a private function so we can change what it returns at any time.

File: packages/@webex/plugin-meetings/test/unit/spec/reachability/index.ts [hunk 10]
```diff
@@ -1033,3 +1762,309 @@ describe('getReachabilityMetrics', () => {
...
+
+  public testSendMetric() {
+    return this.sendMetric();
+  }
+
+  public setFakeClusterReachability(fakeClusterReachability) {
+    this.clusterReachability = fakeClusterReachability;
+  }
+}
+describe('getStatistics', () => {
```
Comment (sreenara) [inline packages/@webex/plugin-meetings/test/unit/spec/reachability/index.ts:1787]: nitpick, please add a new line here

**Review by sreenara — APPROVED** [review]
Just a couple of questions and non-blocking comments. Nice work on the tests!

File: packages/@webex/plugin-meetings/src/reachability/clusterReachability.ts [hunk 11]
```diff
@@ -23,10 +21,32 @@ export type ClusterReachabilityResult = {
...
+  protocol: 'udp' | 'tcp' | 'xtls';
+  clientMediaIPs: string[];
+};
+
+export const Events = {
+  resultReady: 'resultReady', // emitted when a cluster is reached successfully using specific protocol
+  clientMediaIpsUpdated: 'clientMediaIpsUpdated', // emitted when more public IPs are found after resultReady was already sent for a given protocol
+} as const;
+
+export type Events = Enum<typeof Events>;
```
[the remark this answers is not in the document]
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/reachability/clusterReachability.ts:43]: yes, it's explained here: https://confluence-eng-gpk2.cisco.com/conf/display/WTWC/How+to+use+enums+and+alternatives

File: packages/@webex/plugin-meetings/src/reachability/index.ts [hunk 12]
```diff
@@ -60,24 +67,39 @@ export type ReachabilityResults = Record<
   }
 >;
 
+// timeouts in seconds
+const DEFAULT_TIMEOUT = 3;
+const VIDEO_MESH_TIMEOUT = 1;
+const OVERALL_TIMEOUT = 15;
```
[the remark this answers is not in the document]
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/reachability/index.ts:73]: yes, I'm starting with 15s as I want to gather the metrics to see the distribution of latencies and what are the longest ones we get. For now the app won't do anything on the overall timeout anyway, only after gathering those metrics we'll decide on the actual best value for that overall timeout. I've updated the wiki to say that.

File: packages/@webex/plugin-meetings/src/reachability/index.ts [hunk 13]
```diff
@@ -395,16 +409,207 @@ export default class Reachability {
...
+    }, OVERALL_TIMEOUT * 1000);
+  }
+
+  /**
+   * Stores given reachability results in local storage
+   *
+   * @param {ReachabilityResults} results
+   * @returns {Promise<void>}
+   */
+  private async storeResults(results: ReachabilityResults) {
```
[the remark this answers is not in the document]
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/reachability/index.ts:575]: SDK has always been storing it in local storage, I haven't changed it. I guess the idea was to have results ready to use in case we reload the page and don't do reachability before joining a meeting, but right now reachability is always done as part of `webex.meetings.register()` call that you have to make before being able to join anything, so there is not much benefit of using local storage. We might want to re-evaluate this, but it's out of scope for this PR.

File: packages/@webex/plugin-meetings/test/unit/spec/reachability/index.ts (hunk 10 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/test/unit/spec/reachability/index.ts:1788]: done
